### PR TITLE
I made new mergesort

### PR DESCRIPTION
**Pull request by cnuhoya**
this is merge sort 
Good performance

### CODE REVIEW

**Comment by varunu28** [comment]
How is it different from [this](https://github.com/TheAlgorithms/Java/blob/master/Sorts/MergeSort.java) ?

**Comment by cnuhoya** [comment]
Loop invariant project : Statement that should be maintained during loop

File: Sorts/(new)mergesort.java [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+         while (array1index < array1.length && array2index < array2.length ) {
+         	if ( array1[array1index] < array2[array2index]) {
+             	array3[array3index] = array1[array1index];
+                     array1index++; } 
+             else {
+             	array3[array3index] = array2[array2index];
+                     array2index++; }
+             array3index++; 
+         }
+         if(array1index < array1.length) {
```
Comment (christianbender) [inline Sorts/(new)mergesort.java]: You don't need this if-structure. The while loop will do the job.

File: Sorts/(new)mergesort.java [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+             while (array1index < array1.length) {
+             	array3[array3index] = array1[array1index];
+             	array1index++;
+                     array3index++;
+             }
+         }
+         if(array2index < array2.length) {
```
Comment (christianbender) [inline Sorts/(new)mergesort.java]: The same thing as above. The while loop ```while (array2index < array2.length)``` will do the job.

File: Sorts/(new)mergesort.java [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+                     array3index++;
+             }
+         }
+}
+
+public static void merge_sort (int[] array) {
+        int n = array.length;       
+        if (n == 1) return;       
+        int[] array1 = new int[n/2];
+        int[] array2 = new int[n - n/2];
```
Comment (christianbender) [inline Sorts/(new)mergesort.java]: Instead of ```n - n/2``` you can write ```n/2```

File: Sorts/(new)mergesort.java [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+             	array3[array3index] = array2[array2index];
+             	array2index++;
+                     array3index++;
+             }
+         }
+}
+
+public static void merge_sort (int[] array) {
+        int n = array.length;       
+        if (n == 1) return;
```
Comment (christianbender) [inline Sorts/(new)mergesort.java]: Instead of ```n == 1``` you can write ```n <= 1```. For the case n = 0

File: Sorts/(new)mergesort.java [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+
+public static void merge_sort (int[] array) {
+        int n = array.length;       
+        if (n == 1) return;       
+        int[] array1 = new int[n/2];
+        int[] array2 = new int[n - n/2];
+        for (int i = 0; i< n/2; i++) {
+                array1[i] = array[i];
+        }
+        for (int i = 0; i< n - n/2; i++) {
```
Comment (christianbender) [inline Sorts/(new)mergesort.java]: Instead of ```n - n/2``` you can write ```n/2```

File: Sorts/(new)mergesort.java (hunk 4 above)
Comment (christianbender) [inline Sorts/(new)mergesort.java]: Use curly braces! See http://www.oracle.com/technetwork/java/javase/documentation/codeconventions-142311.html#449

File: Sorts/(new)mergesort.java [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+                array2[i] = array[i + n/2];
+        }       
+        merge_sort(array1);
+        merge_sort(array2);      
+        merge(array1, array2, array);
+}
+	public static void main(String[] args) throws IOException {
+		// TODO Auto-generated method stub
+		FileReader reader = new FileReader("C:\\Users\\CSE\\Desktop\\homwork\\input.txt");
```
Comment (christianbender) [inline Sorts/(new)mergesort.java]: Correct the indent

File: Sorts/(new)mergesort.java [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+    	BufferedReader rd = new BufferedReader(reader);
+    	  int index = 0;
```
Comment (christianbender) [inline Sorts/(new)mergesort.java]: Correct the indent in the full main-method.

File: Sorts/(new)mergesort.java [hunk 8]
```diff
@@ -0,0 +1,99 @@
...
+    	  int maxsize;
+    	  String line = null;
+    	  String getLine = "";
+    	  while ((line = rd.readLine()) != null) {
+    		   System.out.println(line);
+    		   getLine = getLine + line;
+    		   // �о ��� String�� �����Ų��
+    		  }
```
Comment (christianbender) [inline Sorts/(new)mergesort.java]: Indent the curly brace according to the while-loop.

File: Sorts/(new)mergesort.java (hunk 6 above)
Comment (christianbender) [inline Sorts/(new)mergesort.java]: Please use try-catch for your file access.

File: Sorts/(new)mergesort.java [hunk 9]
```diff
@@ -0,0 +1,99 @@
...
+		FileReader reader = new FileReader("C:\\Users\\CSE\\Desktop\\homwork\\input.txt");
+    	BufferedReader rd = new BufferedReader(reader);
+    	  int index = 0;
+    	  int maxsize;
+    	  String line = null;
+    	  String getLine = "";
+    	  while ((line = rd.readLine()) != null) {
+    		   System.out.println(line);
+    		   getLine = getLine + line;
+    		   // �о ��� String�� �����Ų��
```
Comment (christianbender) [inline Sorts/(new)mergesort.java]: Please use english for your comments and user information.

File: Sorts/(new)mergesort.java [hunk 10]
```diff
@@ -0,0 +1,99 @@
...
+    		  array[index]=Integer.parseInt(st.nextToken());
+    		   // ©���� ������ �迭�� ����
+    		  index++;
+    		  }  
+    		  long start = System.currentTimeMillis(); //�����ϴ� ��� ���
+    		  merge_sort(array);
+    		  long end = System.currentTimeMillis(); //��α׷��� ������ ��� ���
+    		  System.out.println( "���� �ð� : " + ( end - start )/1000.0 +"��");
+    		  String st_list = "";
+    		  FileWriter writer = new FileWriter("C:\\Users\\CSE\\Desktop\\homwork\\201302491_output.txt");
```
Comment (christianbender) [inline Sorts/(new)mergesort.java]: Please use try-catch for your file access.

File: Sorts/(new)mergesort.java [hunk 11]
```diff
@@ -0,0 +1,99 @@
+package algorithm_homework;
+
```
Comment (christianbender) [inline Sorts/(new)mergesort.java]: Please add comments (in english) in your code.

File: Sorts/(new)mergesort.java [hunk 12]
```diff
@@ -0,0 +1,99 @@
...
+    	 int array1index = 0;
+         int array2index = 0;
+         int array3index = 0;     
+         while (array1index < array1.length && array2index < array2.length ) {
+         	if ( array1[array1index] < array2[array2index]) {
+             	array3[array3index] = array1[array1index];
+                     array1index++; } 
+             else {
+             	array3[array3index] = array2[array2index];
```
Comment (christianbender) [inline Sorts/(new)mergesort.java]: Please correct the indent of the else-block

File: Sorts/(new)mergesort.java [hunk 13]
```diff
@@ -0,0 +1,99 @@
...
+import java.io.BufferedReader;
+import java.io.FileNotFoundException;
+import java.io.FileReader;
+import java.io.FileWriter;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.StringTokenizer;
+
+public class mergesort {
```
Comment (christianbender) [inline Sorts/(new)mergesort.java]: In java class-names begins with upper case. Change the file name according to the class name!

File: Sorts/(new)mergesort.java [hunk 14]
```diff
@@ -0,0 +1,99 @@
...
+         if(array2index < array2.length) {
+             while (array2index < array2.length) {
+             	array3[array3index] = array2[array2index];
+             	array2index++;
+                     array3index++;
+             }
+         }
+}
+
+public static void merge_sort (int[] array) {
```
Comment (christianbender) [inline Sorts/(new)mergesort.java]: Please use camel case names.

**Pull request closed without merging**